### PR TITLE
Buffs Service Borgs

### PR DESCRIPTION
**Pull request by Kyep**
Service borgs now have the following additional reagent dispensers:
- Orange Juice (cures some viruses, reduces oxy damage slightly)
- Tomato Juice (reduces fire damage ~0.2/second)
- Lime Juice & Tea (each reduces tox damage ~0.2/second)
- Milk (reduces brute damage ~0.2/second)
- Coffee (reduces paralysis/stunned/weakened status 1/second)
- Sugar (intended for addition to coffee/tea)

They also get a health analyser, so they can scan crew members, notice they're injured, and serve them a refreshing drink for RP and minor healing. They use the Rapid Service Fabricator to create a glass, then a reagent dispenser on the glass to fill it. These dispensers recharge from the cyborg's battery.

When emagged, they still have their dispenser for poisoned beer, but the poison has also been changed to have the same metabolism rate as cyanide. What this means is that it takes 15 times as long to deplete as it used to. Essentially, if someone drinks a significant amount of poisoned beer, and doesn't get dialysis or similar soon, they die.

Finally, their RSF can create some snack food items: tofu burgers, tofu snacks, noodles, and chicken soup.

Based on a suggestions forum post: https://nanotrasen.se/forum/topic/11776-give-service-borgs-more-functionality-a-buff-they-need

🆑 Kyep
add: Service borgs are now able to dispense a variety of fruit juices, as well as coffee and tea. Additionally, they can use a health scanner on crew, and the poisoned beer they get from being emagged is much more deadly. They can also create snackfood using their Rapid Service Fabricator.
/🆑

### CODE REVIEW

**Comment by pinatacolada** [comment]
sugar dispenser is going to have an impact on borers, not that i think its a bad thing, just a thing worth mentioning

**Comment by The-Driker** [comment]
I will not approve of this until you add in the ability to dispense mini burgers. I REQUIRE BURGER!

**Comment by FalseIncarnate** [comment]
While not necessarily burgers, a service borg being able to serve snacks might not be too amiss.

Obviously nothing special, but maybe something similar to a vending machine snack like chips or a candy bar?

Not sure they need the health scanner, they should already be able to use the medical HUD (don't think that's tied to the medical module) to get an overall sense of health, without them stepping too heavily on the toes of the medical module.

I'd also remove the lime juice and leave it as tea. No sense having multiple options that have the same result, just unnecessary clutter at that point, unless your intent is to have them replace the bartender, in which case there is lime juice on the dispenser already anyways.

**Comment by Landerlow** [comment]
I love this. Thank you very much. I have one additional request perhaps: Would it be possible to give service borgs the ability to dispense small snacks? Think of one-bite snacks like mini-burgers.

I like the different drinks options. Service borg is mostly a RP-module, this adds to the roleplay. The health scanner is nice but not necessary. 

Would it be possible to perhaps give them the ability to put the shaker or something in drink dispensers like soda fountain, etc? I mean, they won't replace bartender if there is one. But if there ISN'T one, it could be a nice addition. 

I like what you've made so far, I"m very happy about this.

**Comment by Kyep** [comment]
UPDATE:

The Rapid Service Fabricator can now create Tofu Turkey platters, Turkey Burgers, Chicken Soup cartons, and Chinese Newdles takeout cartons. 

These have a hefty (500!) energy cost to create.

**Comment by TDSSS** [comment]
Chicken Soup? The most OP of all snackfoods, that which nourishes the scientists? Like, I haven't checked the code but doesn't it take you from hungry to full, pretty much?

**Comment by Kyep** [comment]
Realistically:
- If the service borg could only serve junk food, then there would never be any reason to get food from them. After all, you can easily get junk food from vendors already. Thus, they *have to* have healthy, filling food to dispense - or there is no point in them having food at all.
- Chefs aren't always available. Nor is botany, upon which they depend. There must be healthy food options that don't revolve around the chef. Service borgs, being not present in most shifts, are one such potential option. Though, the food they serve, while healthy, is bland, like tofu, and is not going to be as appealing as food from the chef.

**Comment by SkeletalElite** [comment]
Also you can just drink hot chocolate and take your food from empty to full that way so food isn't really a problem even on chefless shifts

**Comment by The-Driker** [comment]
>Turkey Burgers
`Turkey` `Burgers`
**_You have sinned against me this day._**

**Comment by Kyep** [comment]
😈

**Comment by Landerlow** [comment]
How about some more functionality with the drink dispensers from the bar?

Also I love you Kyet

**Comment by FalseIncarnate** [comment]
I'd probably remove the tofurkey, replace it with something simpler. That thing is like a 2nd/3rd tier chef recipe (need to make tofu and stuffing, stuffing requiring making bread first), plus contains ether. To be fair, the tofuburger is a 2nd tier by the same requirements: tofu and bun.

Additionally, tofu is plantmatter, not nutriment, so carnivores won't benefit from eating it. Consider either something like a synthburger or maybe a good ol' donut to go with the coffee you are serving.

**Comment by Landerlow** [comment]
Consider making it give items which are in sustenance vendors

File: code/game/objects/items/weapons/RSF.dm [hunk 1]
```diff
@@ -55,6 +55,22 @@ RSF
...
+		return
+	if(mode == 8)
+		mode = 9
+		to_chat(user, "Changed dispensing mode to 'Snack - Tofu Burger'")
+		return
+	if(mode == 9)
+		mode = 10
+		to_chat(user, "Changed dispensing mode to 'Snack - Noodles'")
+		return
+	if(mode == 10)
```
Comment (Crazylemon64) [inline code/game/objects/items/weapons/RSF.dm]: These REALLY look like they could benefit from a set of #defines.

File: code/game/objects/items/weapons/RSF.dm [hunk 2]
```diff
@@ -232,6 +248,115 @@ RSF
 				to_chat(user, "The RSF now holds [matter]/30 fabrication-units.")
 				desc = "A RSF. It currently holds [matter]/30 fabrication-units."
 		return
+	else if(istype(A, /obj/structure/table) && mode == 7)
+		if(istype(A, /obj/structure/table) && matter >= 1)
```
Comment (Crazylemon64) [inline code/game/objects/items/weapons/RSF.dm]: Isn't this `istype` check redundant?

File: code/game/objects/items/weapons/RSF.dm [hunk 3]
```diff
@@ -232,6 +248,115 @@ RSF
...
+				var/mob/living/silicon/robot/engy = user
+				engy.cell.charge -= 500
+			else
+				matter--
+				to_chat(user, "The RSF now holds [matter]/30 fabrication-units.")
+				desc = "A RSF. It currently holds [matter]/30 fabrication-units."
+		return
+
+	else if(istype(A, /turf/simulated/floor) && mode == 7)
+		if(istype(A, /turf/simulated/floor) && matter >= 1)
```
Comment (Crazylemon64) [inline code/game/objects/items/weapons/RSF.dm]: Redundant `istype`

File: code/game/objects/items/weapons/RSF.dm [hunk 4]
```diff
@@ -232,6 +248,115 @@ RSF
...
+			if(isrobot(user))
+				var/mob/living/silicon/robot/engy = user
+				engy.cell.charge -= 500
+			else
+				matter--
+				to_chat(user, "The RSF now holds [matter]/30 fabrication-units.")
+				desc = "A RSF. It currently holds [matter]/30 fabrication-units."
+		return
+	else if(istype(A, /obj/structure/table) && mode == 8)
+		if(istype(A, /obj/structure/table) && matter >= 1)
```
Comment (Crazylemon64) [inline code/game/objects/items/weapons/RSF.dm]: redundant `istype`

File: code/game/objects/items/weapons/RSF.dm [hunk 5]
```diff
@@ -232,6 +248,115 @@ RSF
...
+				var/mob/living/silicon/robot/engy = user
+				engy.cell.charge -= 500
+			else
+				matter--
+				to_chat(user, "The RSF now holds [matter]/30 fabrication-units.")
+				desc = "A RSF. It currently holds [matter]/30 fabrication-units."
+		return
+
+	else if(istype(A, /turf/simulated/floor) && mode == 8)
+		if(istype(A, /turf/simulated/floor) && matter >= 1)
```
Comment (Crazylemon64) [inline code/game/objects/items/weapons/RSF.dm]: redundant `istype`

File: code/game/objects/items/weapons/RSF.dm [hunk 6]
```diff
@@ -232,6 +248,115 @@ RSF
...
+			if(isrobot(user))
+				var/mob/living/silicon/robot/engy = user
+				engy.cell.charge -= 500
+			else
+				matter--
+				to_chat(user, "The RSF now holds [matter]/30 fabrication-units.")
+				desc = "A RSF. It currently holds [matter]/30 fabrication-units."
+		return
+	else if(istype(A, /obj/structure/table) && mode == 9)
+		if(istype(A, /obj/structure/table) && matter >= 1)
```
Comment (Crazylemon64) [inline code/game/objects/items/weapons/RSF.dm]: redundant `istype`

File: code/game/objects/items/weapons/RSF.dm [hunk 7]
```diff
@@ -232,6 +248,115 @@ RSF
...
+				var/mob/living/silicon/robot/engy = user
+				engy.cell.charge -= 500
+			else
+				matter--
+				to_chat(user, "The RSF now holds [matter]/30 fabrication-units.")
+				desc = "A RSF. It currently holds [matter]/30 fabrication-units."
+		return
+
+	else if(istype(A, /turf/simulated/floor) && mode == 9)
+		if(istype(A, /turf/simulated/floor) && matter >= 1)
```
Comment (Crazylemon64) [inline code/game/objects/items/weapons/RSF.dm]: redundant `istype`

File: code/game/objects/items/weapons/RSF.dm (hunk 3 above)
Comment (Crazylemon64) [reply code/game/objects/items/weapons/RSF.dm]: the structure of these is really mixed up.
Shouldn't you first check if `A` is the correct type of thing to dispense onto, then check which of the modes you are using?

As it stands, there's a LOT of duplicated code here

**Comment by Kyep** [comment]
Substantial update!

Changes:
- RSF.dm has been refactored to eliminate copypaste and magic numbers.
- The tofu-turkey has been replaced with a donut.
- Several new reagents have been added:
-- Ice (enables iced tea & coffee, which have the same stats as normal versions)
-- Cream (enables doctor's delight, a -1 tox/sec fruit drink)
-- Cola & soda (enables brown star, a fruit drink, as well as plain old cola/sodawater)
-- Tequila/Vodka/Whiskey (which enable tequilla sunrise, margarita, bloody mary, screwdriver, 'syndicate bomb', whiskey cola, whiskey soda).

**Comment by Landerlow** [comment]
Merge pls :))))

**Comment by Spartan6** [comment]
Does the donut include the sprinkled donut, as that actually heals security (if I remember correctly).

**Comment by Kyep** [comment]
All /obj/item/weapon/reagent_containers/food/snacks/donut items have a 30% chance to have sprinkles. It is random.

**Comment by Landerlow** [comment]
I need this AAAAH.

**Comment by ktoma36** [comment]
Now all we need are service borg sprites that aren't derpy :3

**Comment by Landerlow** [comment]
Can we merge this first though :D

File: code/modules/mob/living/silicon/robot/robot_modules.dm [hunk 8]
```diff
@@ -270,10 +270,27 @@
...
+	modules += new /obj/item/weapon/reagent_containers/food/drinks/coffee(src) // -1 paralysis stunned & weakened/sec
+	modules += new /obj/item/weapon/reagent_containers/food/drinks/tea(src)
+	modules += new /obj/item/weapon/reagent_containers/food/drinks/bottle/milk(src) // -0.2 brute/sec
+	modules += new /obj/item/weapon/reagent_containers/food/condiment/sugar(src)
+	modules += new /obj/item/weapon/reagent_containers/food/drinks/ice(src)
+	modules += new /obj/item/weapon/reagent_containers/food/drinks/bottle/cream(src)
+
+	modules += new /obj/item/weapon/reagent_containers/food/drinks/bottle/tequila(src)
+	modules += new /obj/item/weapon/reagent_containers/food/drinks/bottle/vodka(src)
+	modules += new /obj/item/weapon/reagent_containers/food/drinks/bottle/whiskey(src)
```
Comment (Crazylemon64) [inline code/modules/mob/living/silicon/robot/robot_modules.dm]: Ah, how will all these interact with borg rechargers?

Comment (Kyep) [reply code/modules/mob/living/silicon/robot/robot_modules.dm]: Not sure what you mean? They drain your cell to recharge, so they're effectively infinite so long as you recharge occasionally. Is that what you meant?

Comment (FalseIncarnate) [reply code/modules/mob/living/silicon/robot/robot_modules.dm]: You currently do not handle refilling these anywhere is what he is really pointing out. The enzyme and emagged beer bottle are being specially handled in the respawn_consumable proc below (oddly the normal beer bottle isn't). When these bottles run out, they won't refill from a sit in a recharger.

Interestingly, the rechargers seem to be calling these same bits of code it its process proc and then calling respawn_consumables (effectively doubling up the refilling).

Comment (Kyep) [reply code/modules/mob/living/silicon/robot/robot_modules.dm]: When I tested, they seemed to refill automatically. Will retest and be sure.

Comment (Kyep) [reply code/modules/mob/living/silicon/robot/robot_modules.dm]: I retested it to be sure. The reagent containers auto-refill already, without any further additions.

**Comment by Kyep** [comment]
Yeah, I did the test. The reagent containers work fine. The only weird thing is that if you use harm intent with a vodka bottle and hit someone, you smash the built-in bottle over their head and then you don't have the bottle anymore. However, I think that's fine, because its realistic - if your service borg comes with a bottle of booze and a chem maker, they can use the bottle in self-defense, but doing so means they can't serve from it anymore.

File: code/modules/mob/living/silicon/robot/robot_modules.dm (hunk 8 above)
Comment (Crazylemon64) [reply code/modules/mob/living/silicon/robot/robot_modules.dm]: https://github.com/ParadiseSS13/Paradise/blob/master/code/modules/food_and_drinks/drinks/drinks.dm#L98-L109
:fire: :fire: :eyes: :fire: :fire:

**Comment by taukausanake** [comment]
@Kyep Can the borg shank people with the broken bottle then?

**Comment by Kyep** [comment]
Yes, but only once per bottle. They do not get an infinite supply of bottles.

**Comment by The-Driker** [comment]
**NOT SEEING ANY BURGERS ON THIS SPAWN LIST YET!**

**Comment by Landerlow** [comment]
Burgers should not be added as it is food that gives lots of filling and would take a chef time to make. It would be counterproductive and make the chef even more redundant than they already are.

**Comment by The-Driker** [comment]
There are different types of burger. Some of which can't be cooked if I recall correctly. Not sure why those couldn't be adjusted to have lower filling.

**Comment by Landerlow** [comment]
IMO I think food should have lower filling in general = people need to eat more = chef needs to do more.

This would include food from vendors to prevent cheesing.

**Comment by FalseIncarnate** [comment]
Food from vendors already is considered junk food and will make you hungrier faster (and actually has a cut-off where you will refuse to eat more junk food unless starving), but that is all far from the scope of this PR in the first place. Please remain on topic to the changes contained within this PR.

**Comment by Landerlow** [comment]
I'm waiting for it to get merged, it's taking its time

**Comment by adamkad1** [comment]
well i usually cram a lot of stuff onto a plate or something, low cost high reward (highly filling) food. you can actually make a monster worth of food with a plate and lots of icecream. or use a bun and make a screen size burger, because why not

**Comment by tigercat2000** [comment]
LGTM

**Comment by taukausanake** [comment]
They're called Bluespace burgers :^)

**Comment by Landerlow** [comment]
@Kyep  Does a broken bottle refill when you recharge in a recharger, once you've lost the bottle after having smashed it?

**Comment by Kyep** [comment]
@Landerlow I don't think so, I think it just deletes the item. You already get auto-refilling reagents. You can't have infinite bottles :P

**Comment by The-Driker** [comment]
It costs less to create metal walls. There are times when you change things for the sake of other people's mechanics and a time to say “would it really work like this?” At this cost it's not worth using this function. Even with a bluespace cell you could make about 7 low nutriment snacks and then have to sit in a recharger, and I don't even want to HEAR about slime cells, especially now that they are nerfed again.

**Comment by Kyep** [comment]
@The-Driker The heads/maints requested a much higher cost to create snacks, for fear of people spamming them. I have to admit, they may have a point, especially with how much people spammed the cookie dispenser on peacekeeper borgs. Though peacekeeper borgs were a meme and thus lent themselves to this behavior, I can see how people would be concerned about snack spam. So, I am fine with setting the energy cost for snacks really, really high. We can always reduce it later if it turns out to be really too high. For the moment though its probably a good idea to introduce it as something that definitely cannot be spammed.

**Comment by Landerlow** [comment]
Please merge

**Comment by Landerlow** [comment]
Bump?

**Comment by Landerlow** [comment]
...

**Comment by davidchan13** [comment]
Can't we compromise and just give them snack food by default and have proper burgers/sliders, ice cream or what have you in the form of Cyborg Upgrade Modules from robotics?

**Comment by Dumbdumn5** [comment]
:+1:

**Comment by Kyep** [comment]
@davidchan13 Giving them only snackfood would be pointless, as you can already get that stuff more easily from vendors. They *have to* have some healthy options, or people have zero incentive to ever actually eat their food. Cyborg upgrade modules from Robotics are an interesting idea, but not really in scope here. I want to get these basic changes merged before I consider any other changes to them (separate PR).

**Comment by tigercat2000** [comment]
We can tweak this later if the costs are too expensive, or if they're too cheap for that matter. The point of the high costs is so that the borg doesn't make the chef pointless.
👍

**Comment by davidchan13** [comment]
@Kyep I get what you're saying but I'm trying to use the baseline mining borg as example. It's default loadout is enough to get by but the real power of the borg comes from the potential upgrades it can get once mining/science has done their jobs.

Having 'snack' level food on demand is still decently useful on its own (producing soup or raisins, for example) as jobs that are notoriously isolated could just ask the AI to send the borg their way to drop them off a morsel to get them through the end of their project. 

As an avid borg player I believe that borg player I go by the philosophy that a borg should supplement the role it's working in and able to fill in for short term/emergency work if no actual crew can do the job, but it should never be a full replacement.

**Pull request merged**